File: 2015/day2/day2.ml
let lines =
  In_channel.with_open_bin "input.txt" In_channel.input_all
  |> String.split_on_char '\n'
in
List.fold_left
  (fun acc line ->
    let dims = String.split_on_char 'x' line in
    let nums = List.map int_of_string dims in
    match nums with
    | l :: w :: h :: _ ->
        (* let a = l * w in
        let b = l * h in
        let c = w * h in *)
        let pa = (l * 2) + (w * 2) in
        let pb = (l * 2) + (h * 2) in
        let pc = (w * 2) + (h * 2) in
        (* acc + ((a + b + c) * 2) + min a (min b c) *)
        acc + (l * w * h) + min pa (min pb pc)
    | _ -> acc)
  0 lines
|> print_int

File: 2016/day2/day2.py
with open("input.txt") as f:
    lines = f.readlines()

keys = {
    (2, 0): "1",
    (1, 1): "2",
    (2, 1): "3",
    (3, 1): "4",
    (0, 2): "5",
    (1, 2): "6",
    (2, 2): "7",
    (3, 2): "8",
    (4, 2): "9",
    (1, 3): "A",
    (2, 3): "B",
    (3, 3): "C",
    (2, 4): "D",
}

pos = (1, 1)

code = []

for line in lines:
    for char in line.strip():
        x, y = pos
        new = {"U": (x, y - 1), "D": (x, y + 1), "L": (x - 1, y), "R": (x + 1, y)}[char]
        if new in keys:
            pos = new

    code.append(keys[pos])

print("".join(code))
